RenderObject::setNeedsBoundariesUpdate() should avoid work in non-SVG elements

rniwa opened the pull request:
#### e6d4d4d2b7504224bf106c2734c82c18bc8a27b7
<pre>
RenderObject::setNeedsBoundariesUpdate() should avoid work in non-SVG elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=271010">https://bugs.webkit.org/show_bug.cgi?id=271010</a>

Reviewed by Simon Fraser.

This PR refactors RenderObject::setNeedsBoundariesUpdate() into two functions
invalidateCachedBoundaries which invalidates the ancestor with boundary box cache,
and setNeedsBoundariesUpdate, which sets a dirty bit on this boundary box cache.

This avoids the overhead of recursive function calls in setNeedsBoundariesUpdate(),
and avoids traversing upwards of non-SVG render objects. Each render SVG objects
which has a non-trivial setNeedsBoundariesUpdate now sets UsesBoundaryCaching flag
in type specific flags (ReplacedFlag or SVGModelObjectFlag).

A number of call sites of RenderObject::setNeedsBoundariesUpdate now need to call
invalidateCachedBoundaries on its parent since invalidateCachedBoundaries stops at
the first render SVG object which uses boundary box cache.

This PR also removes the unused RenderObject::needsBoundariesUpdate and its overrides.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::willBeRemovedFromTree):
(WebCore::RenderObject::setNeedsBoundariesUpdate):
(WebCore::RenderObject::invalidateCachedBoundaries): Inline the code to walk up
the ancestor objects here along with an early exit for when we&apos;re out of SVG subtree.

* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isSVGRenderer const): Added.
(WebCore::RenderObject::setNeedsTransformUpdate):
(WebCore::RenderObject::needsBoundariesUpdate): Deleted.
(WebCore::RenderObject): Shuffled type flags so that TypeSpecificFlags can be now
16-bits instead of 8-bit like it used to be prior to this PR.
(WebCore::RenderObject::usesBoundaryCaching const): Added.

* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
(WebCore::RenderSVGBlock::styleDidChange):

* Source/WebCore/rendering/svg/RenderSVGInline.cpp:
(WebCore::RenderSVGInline::styleDidChange):

* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::layout):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::LegacyRenderSVGContainer):
(WebCore::LegacyRenderSVGContainer::layout):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.h:
(WebCore::LegacyRenderSVGContainer::needsBoundariesUpdate): Deleted.

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGForeignObject.cpp:
(WebCore::LegacyRenderSVGForeignObject::layout):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
(WebCore::LegacyRenderSVGImage::LegacyRenderSVGImage):
(WebCore::LegacyRenderSVGImage::layout):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.h:
(WebCore::LegacyRenderSVGImage::needsBoundariesUpdate): Deleted.

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::LegacyRenderSVGModelObject):
(WebCore::LegacyRenderSVGModelObject::styleDidChange):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceContainer.cpp:
(WebCore::LegacyRenderSVGResourceContainer::markClientForInvalidation):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::LegacyRenderSVGRoot):
(WebCore::LegacyRenderSVGRoot::styleDidChange):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.h:
(WebCore::LegacyRenderSVGRoot::needsBoundariesUpdate): Deleted.

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::LegacyRenderSVGShape):
(WebCore::LegacyRenderSVGShape::layout):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:
(WebCore::LegacyRenderSVGShape::needsBoundariesUpdate): Deleted.

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGViewportContainer.cpp:
(WebCore::LegacyRenderSVGViewportContainer::calcViewport):

Canonical link: <a href="https://commits.webkit.org/276201@main">https://commits.webkit.org/276201@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c3e478776a95df673eb665f832c3c0579bbe1aef

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43885 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22931 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46308 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46523 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39962 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46188 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26814 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20329 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36208 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44459 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19955 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37788 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17210 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17477 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38886 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1935 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40046 "Found 1 new API test failure: TestWebKitAPI.DataDetectorTests.LoadWKWebViewWithDataDetectorTypePhoneNumber (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39168 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48085 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18889 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15466 "Found 1 new test failure: http/tests/navigation/parsed-iframe-dynamic-form-back-entry.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43022 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20282 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41725 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20486 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6023 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19906 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -251,6 +251,7 @@ class RenderObject : public CachedImageClient, public CanMakeCheckedPtr {
         IsMedia = 1 << 1,
         IsWidget = 1 << 2,
         IsViewTransitionCapture = 1 << 3,
+        HasNeedsBoundariesUpdate = 1 << 5,
```
The name of this flag is confusing, which makes this code above confusing:
```
if (renderer->hasNeedsBoundariesUpdate()) {
  renderer->setNeedsBoundariesUpdate();
  ...
```

The flag and `hasNeedsBoundariesUpdate` should be something like `cachesBoundingBoxes` or something.

rniwa committed:
RenderObject::setNeedsBoundariesUpdate() should avoid work in non-SVG elements
https://bugs.webkit.org/show_bug.cgi?id=271010

Reviewed by Simon Fraser.

This PR refactors RenderObject::setNeedsBoundariesUpdate() into two functions
invalidateCachedBoundaries which invalidates the ancestor with boundary box cache,
and setNeedsBoundariesUpdate, which sets a dirty bit on this boundary box cache.

This avoids the overhead of recursive function calls in setNeedsBoundariesUpdate(),
and avoids traversing upwards of non-SVG render objects. Each render SVG objects
which has a non-trivial setNeedsBoundariesUpdate now sets UsesBoundaryCaching flag
in type specific flags (ReplacedFlag or SVGModelObjectFlag).

A number of call sites of RenderObject::setNeedsBoundariesUpdate now need to call
invalidateCachedBoundaries on its parent since invalidateCachedBoundaries stops at
the first render SVG object which uses boundary box cache.

This PR also removes the unused RenderObject::needsBoundariesUpdate and its overrides.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::willBeRemovedFromTree):
(WebCore::RenderObject::setNeedsBoundariesUpdate):
(WebCore::RenderObject::invalidateCachedBoundaries): Inline the code to walk up
the ancestor objects here along with an early exit for when we're out of SVG subtree.

* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isSVGRenderer const): Added.
(WebCore::RenderObject::setNeedsTransformUpdate):
(WebCore::RenderObject::needsBoundariesUpdate): Deleted.
(WebCore::RenderObject): Shuffled type flags so that TypeSpecificFlags can be now
16-bits instead of 8-bit like it used to be prior to this PR.
(WebCore::RenderObject::usesBoundaryCaching const): Added.

* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
(WebCore::RenderSVGBlock::styleDidChange):

* Source/WebCore/rendering/svg/RenderSVGInline.cpp:
(WebCore::RenderSVGInline::styleDidChange):

* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::layout):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::LegacyRenderSVGContainer):
(WebCore::LegacyRenderSVGContainer::layout):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.h:
(WebCore::LegacyRenderSVGContainer::needsBoundariesUpdate): Deleted.

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGForeignObject.cpp:
(WebCore::LegacyRenderSVGForeignObject::layout):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
(WebCore::LegacyRenderSVGImage::LegacyRenderSVGImage):
(WebCore::LegacyRenderSVGImage::layout):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.h:
(WebCore::LegacyRenderSVGImage::needsBoundariesUpdate): Deleted.

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::LegacyRenderSVGModelObject):
(WebCore::LegacyRenderSVGModelObject::styleDidChange):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceContainer.cpp:
(WebCore::LegacyRenderSVGResourceContainer::markClientForInvalidation):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::LegacyRenderSVGRoot):
(WebCore::LegacyRenderSVGRoot::styleDidChange):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.h:
(WebCore::LegacyRenderSVGRoot::needsBoundariesUpdate): Deleted.

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::LegacyRenderSVGShape):
(WebCore::LegacyRenderSVGShape::layout):

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:
(WebCore::LegacyRenderSVGShape::needsBoundariesUpdate): Deleted.

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGViewportContainer.cpp:
(WebCore::LegacyRenderSVGViewportContainer::calcViewport):

Canonical link: https://commits.webkit.org/276201@main
